File: server/FairwayHq.Api/Authorization/RolePermissions.cs
using static FairwayHq.Api.Authorization.Permissions;

namespace FairwayHq.Api.Authorization;

/// <summary>
/// The canonical role → permissions matrix. This is the source of truth
/// the <see cref="PermissionClaimsTransformation"/> uses to expand a
/// user's role claims into permission claims.
///
/// Keep in sync with the matrix in docs/decisions/0003-auth-rbac.md.
/// </summary>
public static class RolePermissions
{
    public static readonly IReadOnlyDictionary<string, IReadOnlySet<string>> Map =
        new Dictionary<string, IReadOnlySet<string>>
        {
            [Roles.Owner] = new HashSet<string>(Permissions.All),

            [Roles.Manager] = new HashSet<string>
            {
                TeeTimesRead, TeeTimesWrite, TeeTimesCheckin, TeeTimesCancel,
                MembersRead, MembersWrite, MembersSuspend,
                MembersApplicationsRead, MembersApplicationsWrite, MembersOverviewRead,
                LedgerRead, LedgerCharge, LedgerPayment, LedgerVoid, DunningRun,
                CoursesRead, CoursesWrite, NinesRead, NinesWrite,
                StaffRead, StaffWrite, ShiftsRead, ShiftsWrite, TemplatesRead, TemplatesWrite,
                ProductsRead, ProductsWrite, ProductsStock,
                TournamentsRead, TournamentsWrite,
                MaintenanceRead, MaintenanceWrite,
                TabsRead, TabsWrite, TabsPayment, TabsSettle, TabsVoid,
                SnapshotExport,
                SystemHealthRead,
                AdminRead,
                // Excludes: import:run, snapshot:restore, system:clear,
                // admin:write, admin:host. Destructive / system-owner only.
            },

            [Roles.Pro] = new HashSet<string>
            {
                TeeTimesRead, TeeTimesWrite, TeeTimesCheckin, TeeTimesCancel,
                MembersRead, MembersWrite,
                MembersApplicationsRead,
                MembersOverviewRead,
                LedgerRead,
                CoursesRead,
                NinesRead,
                ProductsRead, ProductsStock,
                TournamentsRead, TournamentsWrite,
                MaintenanceRead,
                TabsRead, TabsWrite, TabsPayment, TabsSettle,
            },

            [Roles.AssistantPro] = new HashSet<string>
            {
                TeeTimesRead, TeeTimesWrite, TeeTimesCheckin,
                MembersRead,
                MembersOverviewRead,
                CoursesRead,
                NinesRead,
                ProductsRead,
                TournamentsRead,
                TabsRead, TabsWrite, TabsPayment, TabsSettle,
            },

            [Roles.ProShop] = new HashSet<string>
            {
                TeeTimesRead, TeeTimesCheckin,
                MembersRead,
                MembersOverviewRead,
                LedgerPayment,
                CoursesRead,
                ProductsRead, ProductsStock,
                TabsRead, TabsWrite, TabsPayment, TabsSettle,
            },

            [Roles.Greenkeeper] = new HashSet<string>
            {
                TeeTimesRead,
                CoursesRead,
                NinesRead,
                MaintenanceRead, MaintenanceWrite,
            },

            [Roles.Starter] = new HashSet<string>
            {
                TeeTimesRead, TeeTimesCheckin,
            },
        };

    /// <summary>
    /// Expand a set of role names into the union of their permissions.
    /// Unknown role names are silently ignored — the security boundary
    /// is "user gets only what mapped roles grant," not "block on
    /// unrecognized role." Keeps the app forward-compatible if Keycloak
    /// introduces additional roles before the app catches up.
    /// </summary>
    public static IReadOnlySet<string> PermissionsFor(IEnumerable<string> roles)
    {
        var result = new HashSet<string>();
        foreach (var role in roles)
        {
            if (Map.TryGetValue(role, out var perms))
            {
                result.UnionWith(perms);
            }
        }
        return result;
    }
}

File: client/src/auth/rolePermissions.ts
// Mirrors server/FairwayHq.Api/Authorization/*.cs. Keep in sync.
//
// The canonical role → permissions matrix duplicated from the server's
// `RolePermissions.cs`. The server is the security boundary; this copy
// drives UX-level role-aware rendering only. Drift is allowed; security
// is preserved as long as the server's matrix stays canonical.

import {
  ADMIN_READ,
  ALL_PERMISSIONS,
  COURSES_READ,
  COURSES_WRITE,
  DUNNING_RUN,
  LEDGER_CHARGE,
  LEDGER_PAYMENT,
  LEDGER_READ,
  LEDGER_VOID,
  MAINTENANCE_READ,
  MAINTENANCE_WRITE,
  MEMBERS_APPLICATIONS_READ,
  MEMBERS_APPLICATIONS_WRITE,
  MEMBERS_OVERVIEW_READ,
  MEMBERS_READ,
  MEMBERS_SUSPEND,
  MEMBERS_WRITE,
  NINES_READ,
  NINES_WRITE,
  PRODUCTS_READ,
  PRODUCTS_STOCK,
  PRODUCTS_WRITE,
  SHIFTS_READ,
  SHIFTS_WRITE,
  SNAPSHOT_EXPORT,
  STAFF_READ,
  STAFF_WRITE,
  SYSTEM_HEALTH_READ,
  TABS_PAYMENT,
  TABS_READ,
  TABS_SETTLE,
  TABS_VOID,
  TABS_WRITE,
  TEE_TIMES_CANCEL,
  TEE_TIMES_CHECKIN,
  TEE_TIMES_READ,
  TEE_TIMES_WRITE,
  TEMPLATES_READ,
  TEMPLATES_WRITE,
  TOURNAMENTS_READ,
  TOURNAMENTS_WRITE,
} from "./permissions";
import {
  ASSISTANT_PRO,
  GREENKEEPER,
  MANAGER,
  OWNER,
  PRO,
  PRO_SHOP,
  STARTER,
} from "./roles";

/**
 * Role → permissions map. Exact translation of server `RolePermissions.Map`.
 * Owner gets every permission; everyone else is an explicit allow-list.
 */
export const ROLE_PERMISSIONS: ReadonlyMap<string, ReadonlySet<string>> =
  new Map<string, ReadonlySet<string>>([
    [OWNER, new Set<string>(ALL_PERMISSIONS)],

    [
      MANAGER,
      new Set<string>([
        TEE_TIMES_READ,
        TEE_TIMES_WRITE,
        TEE_TIMES_CHECKIN,
        TEE_TIMES_CANCEL,
        MEMBERS_READ,
        MEMBERS_WRITE,
        MEMBERS_SUSPEND,
        MEMBERS_APPLICATIONS_READ,
        MEMBERS_APPLICATIONS_WRITE,
        MEMBERS_OVERVIEW_READ,
        LEDGER_READ,
        LEDGER_CHARGE,
        LEDGER_PAYMENT,
        LEDGER_VOID,
        DUNNING_RUN,
        COURSES_READ,
        COURSES_WRITE,
        NINES_READ,
        NINES_WRITE,
        STAFF_READ,
        STAFF_WRITE,
        SHIFTS_READ,
        SHIFTS_WRITE,
        TEMPLATES_READ,
        TEMPLATES_WRITE,
        PRODUCTS_READ,
        PRODUCTS_WRITE,
        PRODUCTS_STOCK,
        TOURNAMENTS_READ,
        TOURNAMENTS_WRITE,
        MAINTENANCE_READ,
        MAINTENANCE_WRITE,
        TABS_READ,
        TABS_WRITE,
        TABS_PAYMENT,
        TABS_SETTLE,
        TABS_VOID,
        SNAPSHOT_EXPORT,
        SYSTEM_HEALTH_READ,
        ADMIN_READ,
        // Excludes: import:run, snapshot:restore, system:clear,
        // admin:write, admin:host. Destructive / system-owner only.
      ]),
    ],

    [
      PRO,
      new Set<string>([
        TEE_TIMES_READ,
        TEE_TIMES_WRITE,
        TEE_TIMES_CHECKIN,
        TEE_TIMES_CANCEL,
        MEMBERS_READ,
        MEMBERS_WRITE,
        MEMBERS_APPLICATIONS_READ,
        MEMBERS_OVERVIEW_READ,
        LEDGER_READ,
        COURSES_READ,
        NINES_READ,
        PRODUCTS_READ,
        PRODUCTS_STOCK,
        TOURNAMENTS_READ,
        TOURNAMENTS_WRITE,
        MAINTENANCE_READ,
        TABS_READ,
        TABS_WRITE,
        TABS_PAYMENT,
        TABS_SETTLE,
      ]),
    ],

    [
      ASSISTANT_PRO,
      new Set<string>([
        TEE_TIMES_READ,
        TEE_TIMES_WRITE,
        TEE_TIMES_CHECKIN,
        MEMBERS_READ,
        MEMBERS_OVERVIEW_READ,
        COURSES_READ,
        NINES_READ,
        PRODUCTS_READ,
        TOURNAMENTS_READ,
        TABS_READ,
        TABS_WRITE,
        TABS_PAYMENT,
        TABS_SETTLE,
      ]),
    ],

    [
      PRO_SHOP,
      new Set<string>([
        TEE_TIMES_READ,
        TEE_TIMES_CHECKIN,
        MEMBERS_READ,
        MEMBERS_OVERVIEW_READ,
        LEDGER_PAYMENT,
        COURSES_READ,
        PRODUCTS_READ,
        PRODUCTS_STOCK,
        TABS_READ,
        TABS_WRITE,
        TABS_PAYMENT,
        TABS_SETTLE,
      ]),
    ],

    [
      GREENKEEPER,
      new Set<string>([
        TEE_TIMES_READ,
        COURSES_READ,
        NINES_READ,
        MAINTENANCE_READ,
        MAINTENANCE_WRITE,
      ]),
    ],

    [STARTER, new Set<string>([TEE_TIMES_READ, TEE_TIMES_CHECKIN])],
  ]);

/**
 * Expand a set of role names into the union of their permissions. Unknown
 * role names are silently ignored — the security boundary is "user gets
 * only what mapped roles grant," not "block on unrecognized role." Keeps
 * the app forward-compatible if Keycloak introduces a role before the
 * client's matrix catches up.
 *
 * Analogous to the server's `RolePermissions.PermissionsFor`.
 */
export function permissionsFor(roles: string[]): Set<string> {
  const result = new Set<string>();
  for (const role of roles) {
    const perms = ROLE_PERMISSIONS.get(role);
    if (perms) {
      for (const p of perms) result.add(p);
    }
  }
  return result;
}
